### PR TITLE
Issues/3732 prolific recruiter

### PR DESCRIPTION
**Pull request by jessesnyder**
## Description
Adds support for using the Prolific recruitment platform in Dallinger.

Note: this PR changes the signature of the `Recruiter.reward_bonus()` method. This should be 
regarded as a breaking change, since custom recruiters implemented outside of Dallinger core
will now be incompatible.

## Added dependencies
This PR adds a dependency on the [tenacity library](https://pypi.org/project/tenacity/) for declarative retry configuration.

## How Has This Been Tested?
* automated tests, to the extent possible (Prolific does not offer the same sandbox features as MTurk)
* deployments of the Bartlett1932 demo to Heroku, and testing of a single participant through the full lifecycle, including payment and bonus.

## Disclaimers
* The documentation of Prolific configuration options is decent, but in general the documentation around recruitment could use more work. The docs were originally written at a time when MTurk was the only game in town, and the structure of the docs has this assumption baked in to some degree.
* Documentation of Prolific recruitment subtleties is a lot more spare than I intended. I think the best approach here would be to have someone actually familiar with Prolific and Dallinger work on these, and then if necessary a developer can integrate the content into the Dallinger docs.

### CODE REVIEW

File: dallinger/config.py [hunk 1]
```diff
@@ -72,12 +72,18 @@ def is_valid_json(value):
...
     ("mode", six.text_type, []),
+    ("mturk_qualification_blocklist", six.text_type, ["qualification_blacklist"]),
+    ("mturk_qualification_requirements", six.text_type, [], False, [is_valid_json]),
     ("num_dynos_web", int, []),
     ("num_dynos_worker", int, []),
     ("organization_name", six.text_type, []),
     ("port", int, ["PORT"]),
-    ("mturk_qualification_blocklist", six.text_type, ["qualification_blacklist"]),
-    ("mturk_qualification_requirements", six.text_type, [], False, [is_valid_json]),
+    ("prolific_api_token", six.text_type, [], True),
```
Comment (alecpm) [inline dallinger/config.py:81]: I wonder if it at some point we might consider doing this sort of thing as a separate package from Dallinger core, so that it could operate on its own release schedule, etc.

I think we've added enough extensibility to Dallinger for a 3rd party recruiter to register itself, but it would be interesting to find out if that's really the case. Doing this inside of Dallinger core is definitely the path of least resistance though.

File: dallinger/recruiters.py [hunk 2]
```diff
@@ -128,7 +132,7 @@ def exit_response(self, experiment, participant):
         """
         raise NotImplementedError
 
-    def reward_bonus(self, assignment_id, amount, reason):
+    def reward_bonus(self, worker_id, assignment_id, amount, reason):
```
Comment (alecpm) [inline dallinger/recruiters.py:135]: I don't love that we need to make an API/signature change for this. Maybe there's a way we could make the arguments more flexible? I guess that's tricky because this is always called from the worker, would passing the participant object provide a simpler API with more flexibility (though possibly harder to test)?

File: tests/conftest.py [hunk 3]
```diff
@@ -189,7 +237,7 @@ def my_check_output(cmd):
         yield check_output
 
 
-@pytest.fixture(scope="session", autouse=True)
```
Comment (alecpm) [inline tests/conftest.py:192]: Why this change? Was this fixture being applied too broadly?

**Review by alecpm — APPROVED** [review]
The general approach looks good to me, and I don't see any obvious red flags. I like the test setup too. My only concern at this point is that if we are going to change the call signature for `Recruiter.reward_bonus`, I think we should probably do so that allows for more flexibility in the future (e.g. allowing access to all participant properties, including hit id, assignment id, worker id, and all entry information).

File: dallinger/recruiters.py (hunk 2 above)
Comment (jessesnyder) [reply dallinger/recruiters.py:135]: I wrestled with this, too, for the same reasons.  I'm wondering about a `participant_data` parameter, which would just be data (just a `dict` probably) which might get us both benefits. In fact, we already have this in the form of `Participant.json_data()`, except that it mysteriously lacks the DB `id`, but that could be added.

File: tests/conftest.py (hunk 3 above)
Comment (jessesnyder) [reply tests/conftest.py:192]: Any code that used `requests` had some bits several layers down the stack patched with a mock, because `requests` uses netrc. It was even worse than that sounds, because the resulting error was swallowed so I had no idea why my request was failing for quite a while. :woozy_face:

File: dallinger/recruiters.py (hunk 2 above)
Comment (alecpm) [reply dallinger/recruiters.py:135]: I like the idea of using the participant `json_data` as `participant_data` if we're changing the signature. What's the need for the DB `id`? Can't we just look that up using `worker_id` or `assignment_id`?

I'm not sure I understand why we can't just pass the participant object here, since the worker and the recruiter method presumably run in the same transaction/session, but if there's an issue with that, the `json_data` should be relatively future proof, I'd think.

File: tests/conftest.py (hunk 3 above)
Comment (alecpm) [reply tests/conftest.py:192]: ugh

File: dallinger/recruiters.py (hunk 2 above)
Comment (jessesnyder) [reply dallinger/recruiters.py:136]: I wanted to avoid immediately succumbing to the temptation to pass an ORM object when only data is necessary, but we do pass `Participant` objects to some other methods in the `Recruiter` API, so the precedent is already set, for better or worse. What tips the balance: I just discovered that `worker_id` is not in `json_data` either, probably because this is intended to be safely exportable without PII. We definitely need that for rewarding bonuses!

Instead of adding another `[x]_data` method I'll just relent an pass the `Participant`.
